homeassistant/components/nest/__init__.py

```diff
@@ -134,40 +146,72 @@ def nest_update_event_broker(hass, nest):
         hass.async_create_task(hass.config_entries.async_forward_entry_setup(
             entry, component))
 
-    def set_mode(service):
-        """
-        Set the home/away mode for a Nest structure.
+    def set_away_mode(service):
+        """Set the away mode for a Nest structure."""
+        if ATTR_STRUCTURE in service.data:
+            target_structures = service.data[ATTR_STRUCTURE]
+        else:
+            target_structures = hass.data[DATA_NEST].local_structure
+
+        for structure in nest.structures:
+            if structure.name in target_structures:
+                _LOGGER.info("Setting away mode for: %s to: %s",
+                             structure.name, service.data[AWAY_MODE_HOME])
+                structure.away = service.data[AWAY_MODE_HOME]
+
+    def set_eta(service):
+        """Set away mode to away and include ETA for a Nest structure."""
+        if ATTR_STRUCTURE in service.data:
+            target_structures = service.data[ATTR_STRUCTURE]
+        else:
+            target_structures = hass.data[DATA_NEST].local_structure
 
-        You can set optional eta information when set mode to away.
-        """
+        for structure in nest.structures:
+            if structure.name in target_structures and structure.thermostats:
+                _LOGGER.info("Setting mode for: %s to: %s",
+                             structure.name, AWAY_MODE_AWAY)
+                structure.away = AWAY_MODE_AWAY
+
+                now = datetime.utcnow()
+                trip_id = service.data.get(
+                    ATTR_TRIP_ID, "trip_{}".format(int(now.timestamp())))
+                eta_begin = now + service.data[ATTR_ETA]
+                eta_window = service.data.get(ATTR_ETA_WINDOW,
+                                              timedelta(minutes=1))
+                eta_end = eta_begin + eta_window
+                _LOGGER.info("Setting ETA for trip: %s, "
+                             "ETA window starts at: %s and ends at: %s",
+                             trip_id, eta_begin, eta_end)
+                structure.set_eta(trip_id, eta_begin, eta_end)
+            else:
+                _LOGGER.info("No thermostats found in structure: %s, unable "
```

Comment:
This log is still not right. This else clause included two cases, 1) structure is not the target, 2) structure does not have thermostats. I think we should have different logging for it. 1) should be a debug level, 2) should be a info level